src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs

```diff
@@ -118,5 +101,100 @@ private static string FilePathToLockName(string filePath)
+        {
+#if DNXCORE50
+            private static Dictionary<string, SemaphoreWrapper> _nameWrapper =
+                new Dictionary<string, SemaphoreWrapper>();
+                
+            private readonly string _name;
+            private volatile int _refCount = 0;
+#endif
+
+            private readonly Semaphore _semaphore;
+
+            public static SemaphoreWrapper Create(string name, out bool createdNew)
+            {
+                return Create(0, 1, name, out createdNew);
+            }
+            
+            public static SemaphoreWrapper Create(int initialCount, int maximumCount, string name, out bool createdNew)
+            {
+#if DNXCORE50
+                if (RuntimeEnvironmentHelper.IsWindows)
+                {
+                    return new SemaphoreWrapper(new Semaphore(initialCount, maximumCount, name, out createdNew));
+                }
+                else
+                {
+                    var createdNewLocal = false;
+                    SemaphoreWrapper wrapper;
+
+                    lock (_nameWrapper)
+                    {
+                        
+                        if (!_nameWrapper.TryGetValue(name, out wrapper))
+                        {
+                            createdNewLocal = true;
+                            wrapper = new SemaphoreWrapper(new Semaphore(initialCount, maximumCount), name);
+                            _nameWrapper[name] = wrapper;
+                        }
+                        wrapper._refCount++;
+                    }
+
+                    // C# doesn't allow assigning value to an out parameter directly in lambda expression
+                    createdNew = createdNewLocal;
+                    return wrapper;
+                }
+#else
+
+                return new SemaphoreWrapper(new Semaphore(initialCount, maximumCount, name, out createdNew));
+#endif
+            }
+
+            private SemaphoreWrapper(Semaphore semaphore, string name = null)
+            {
+                _semaphore = semaphore;
+#if DNXCORE50
+                _name = name;
+#endif
+            }
+
+            public bool WaitOne(TimeSpan timeout)
+            {
+                return _semaphore.WaitOne(timeout);
+            }
+
+            public int Release()
+            {
+                return _semaphore.Release();
+            }
+
+            public void Dispose()
+            {
+#if DNXCORE50
+                if (RuntimeEnvironmentHelper.IsWindows)
+                {
+                    _semaphore.Dispose();
+                }
+                else
+                {
+                    lock (_nameWrapper)
+                    {
+                        _refCount--;
+                        if (_refCount == 0)
+                        {
+                            _nameWrapper.Remove(_name);
+                            _semaphore.Dispose();
+                        }
+                    }
+
+                }
+#else
+                _semaphore.Dispose();
```

Comment:
can this be done just once at the end for all 3 scenarios? does it need to be inside the lock?